### PR TITLE
feat: replace use of `getMetaMaskIdentities` with `getInternalAccounts`

### PR DESCRIPTION
**Pull request by montelaidev**
## **Description**
Currently, MetaMask faces tight coupling between its UI and the keyring-controller. The UI heavily relies on the keyring-controller's state while also amalgamating information from various sources, such as preferences and balances.

However, this approach presents some challenges. The keyring-controller must be aware of the UI's limitations, like the need for instant account list provision to avoid lag. Moreover, it takes on the responsibility of adding metadata to accounts, such as the associated keyring type, required for displaying account details.

To address these issues, the introduction of the accounts-controller comes into play as a new abstraction layer between the UI and the keyring-controller. This separation of responsibilities allows the keyring-controller to focus on two main tasks:

- Routing requests to the appropriate keyring.
- Persisting the state of the keyrings.

This PR replaces the usage of `getMetaMaskIdentities` with `getInternalAccounts`

Depends On: https://github.com/MetaMask/metamask-extension/pull/21553

## **Related issues**

Resolves https://github.com/MetaMask/accounts-planning/issues/134

## **Manual testing steps**

1. Use a hardware wallet to test all the flows

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Coding Standards](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've clearly explained:
  - [x] What problem this PR is solving.
  - [x] How this problem was solved.
  - [x] How reviewers can test my changes.
- [x] I’ve indicated what issue this PR is linked to: Fixes #???
- [x] I’ve included tests if applicable.
- [x] I’ve documented any added code.
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-extension/blob/develop/.github/guidelines/LABELING_GUIDELINES.md)).
- [x] I’ve properly set the pull request status:
  - [x] In case it's not yet "ready for review", I've set it to "draft".
  - [x] In case it's "ready for review", I've changed it from "draft" to "non-draft".

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by gantunesr** [comment]
I see that this PR also changes the file `ui/pages/snaps/snap-view/snap-settings.js`.

This PR is quite big so the chance of being blocked is high, I strongly recommend opening a new PR with just that change to remove that review dependency.

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.test.js
Comment (gantunesr) [inline ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.test.js:100]: This is quite a repeated mock we use in all tests, we should probably create this `internalAccounts` in a mock object inside the file `test/jest/mocks` and just import it wherever needed. Feel free to do it here or another PR

File: ui/components/app/modals/confirm-remove-account/confirm-remove-account.test.js
Comment (gantunesr) [inline ui/components/app/modals/confirm-remove-account/confirm-remove-account.test.js:78]: Nit: I know this is not from your PR but still
```suggestion
  it('should call method removeAccount with account address', async () => {
```

File: ui/ducks/metamask/metamask.test.js
Comment (gantunesr) [inline ui/ducks/metamask/metamask.test.js:62]: Just a question, is there a reason to have this array in this test and in the others have the methods as a destructure of `EthMethod`?

File: ui/pages/confirmation/stories/success.stories.js
Comment (gantunesr) [inline ui/pages/confirmation/stories/success.stories.js:134]: Here you can see a similar thing to a previous comment were the code is not consistent regarding this mocks

File: ui/pages/permissions-connect/permissions-connect.component.js
Comment (gantunesr) [inline ui/pages/permissions-connect/permissions-connect.component.js:38]: Here you're using `arrayOf` but you're not enforcing the data type of the array items

File: ui/pages/routes/routes.component.test.js
Comment (gantunesr) [inline ui/pages/routes/routes.component.test.js:74]: I think you pushed this by mistake
```suggestion
    it('should render with network change disabled while adding recipient for send flow', () => {
```

File: ui/pages/send/send.test.js
Comment (gantunesr) [inline ui/pages/send/send.test.js:250]: This seems like a odd addition for this PR, were you fixing something unrelated?

**Review by gantunesr — COMMENTED** [review]
Review in progress (45/78 files). I'm going to drop my comments so far

File: ui/components/app/modals/confirm-remove-account/confirm-remove-account.test.js
Comment (montelaidev) [reply ui/components/app/modals/confirm-remove-account/confirm-remove-account.test.js]: Sure, no problem. Changed in 19731a32e04def4e5b573cb76cea4a26451cf118

File: ui/pages/permissions-connect/permissions-connect.component.js
Comment (montelaidev) [reply ui/pages/permissions-connect/permissions-connect.component.js]: Added the shape in 55991ab71addbf1b4c013d0eafbe913870ab97b9

File: ui/pages/routes/routes.component.test.js
Comment (montelaidev) [reply ui/pages/routes/routes.component.test.js]: Fixed in 55991ab71addbf1b4c013d0eafbe913870ab97b9

File: ui/pages/send/send.test.js
Comment (montelaidev) [reply ui/pages/send/send.test.js]: This is an old field, it can be removed now. Changed in 47fd68abbcc340838a25f1a0619cb06d04ceb83b

File: test/e2e/tests/errors.spec.js
Comment (montelaidev) [reply test/e2e/tests/errors.spec.js]: This is no longer needed, removed in bcf77edfb0ba65953109dd1351e7c3102ba02e10

File: ui/components/multichain/account-list-item/account-list-item.js
Comment (k-g-j) [inline ui/components/multichain/account-list-item/account-list-item.js:327]: To simplify things, could we remove the keyring property here and just use `account.metadata.keyring` since the two are the same, and this will be consistent with the account prop type we are using elsewhere. It will require a change to the [mergeAccounts](https://github.com/MetaMask/metamask-extension/blob/642f6d4cf1655edc519de31075d8b14abe0f8e83/ui/components/multichain/account-list-menu/account-list-menu.js#L84) function in the parent component which will also be unnecessary after this PR is merged. This is tech debt for myself since I worked on that component before, I can make a separate PR to update the account-list-menu and remove the unnecessary merging function and props passed down after this one is merged.

File: ui/components/multichain/account-list-menu/account-list-menu.js
Comment (k-g-j) [inline ui/components/multichain/account-list-menu/account-list-menu.js:143]: Since `accounts` will now have all the data from internalAccounts + the cached balance, we wont need to perform a merge anymore.

File: ui/components/multichain/receive-modal/receive-modal.js
Comment (k-g-j) [inline ui/components/multichain/receive-modal/receive-modal.js:34]: It seems like before name could never be undefined, but here `internalAccount?` could be undefined and then we would not have a name.

File: ui/pages/confirmations/components/signature-request-siwe/signature-request-siwe.js
Comment (k-g-j) [inline ui/pages/confirmations/components/signature-request-siwe/signature-request-siwe.js:71]: can we use `[getInternalAccountByAddress](https://github.com/MetaMask/metamask-extension/blob/fix/update-getMetaMaskIdentities-to-getInternalAccounts/ui/selectors/selectors.js#L329)` here?

**Comment by k-g-j** [comment]
I do not have a hardware wallet, but I tested on the E2E test dapp and the SSK all flows and did not notice any bugs or errors in sending or signing.

File: ui/components/app/account-list-item/account-list-item.js
Comment (montelaidev) [reply ui/components/app/account-list-item/account-list-item.js]: Yes, its been removed.

File: ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.test.js
Comment (montelaidev) [reply ui/components/app/edit-gas-fee-popover/edit-gas-fee-popover.test.js]: I'll create a follow on PR to dynamically create the mocks as needed.

File: ui/components/multichain/receive-modal/receive-modal.js
Comment (montelaidev) [reply ui/components/multichain/receive-modal/receive-modal.js]: Yes, it should not be undefined.

File: ui/components/multichain/account-list-item/account-list-item.js
Comment (montelaidev) [reply ui/components/multichain/account-list-item/account-list-item.js:329]: Yes, lets remove this in a separate PR, the mergeAccounts is being using in a few places.

File: ui/ducks/metamask/metamask.test.js
Comment (montelaidev) [reply ui/ducks/metamask/metamask.test.js:62]: It should be migrated to EOA_EVM_METHODS, the keyring api will return user op methods as well.

**Review by gantunesr — COMMENTED** [review]
Left another small comment. The rest of the code looks code to me

edit: I'll approve it once the rest of the comments are resolved

File: ui/components/multichain/account-list-menu/account-list-menu.js
Comment (montelaidev) [reply ui/components/multichain/account-list-menu/account-list-menu.js:143]: To be updated in the mergeAccounts pr

**Comment by gantunesr** [comment]
@montelaidev I think a minor bug in the send flow

https://github.com/MetaMask/metamask-extension/assets/17601467/1d4ad918-9b62-4fb1-b1c1-6c56fd602dfd

**Comment by montelaidev** [comment]
> @montelaidev I found a minor bug in the send flow
> 
>  Screen.Recording.2024-02-15.at.14.32.29.mov

Looks like its the same bug as this https://github.com/MetaMask/metamask-extension/issues/22933, its due to the petnames overriding the names being passed to the component.

**Comment by gantunesr** [comment]
True, let's try to fix it in another PR

**Review by gantunesr — DISMISSED** [review]
lgtm

**Review by k-g-j — DISMISSED** [review]
I reviewed the code changes and tested on a regular EOA and my snap account and all sending and e2e flows worked as I would expect. I changed the nicknames for the accounts and the change persisted in the send flow.
Screenshot 2024-02-16 at 12 47 16 PM
Screenshot 2024-02-16 at 12 47 33 PM

**Comment by plasmacorral** [comment]
Tested on Mac in Sonoma 14.3.1 with Chrome 117.0.5938.92 using Ledger, Lattice, QR, Trezor and Snap accounts.  Also spot checked in Firefox 123.0 Lattice, QR, Trezor and Snap accounts (Ledger not supported >v114)

**Comment by plasmacorral** [comment]
I have encountered a situation where the test dapp can no longer be used to connect an account, after the one that was connected is removed.  It initially struck me as similar to what is reported [here](https://github.com/MetaMask/metamask-mobile/issues/8307), but occurs with connecting an account rather than signatures and I want to be confident its not related to the changes in this PR before we merge.

Reproduction steps:

1. Import SRP
2. Add Hardware wallet (Ledger tested)
3. Connect to [test dapp ](https://metamask.github.io/test-dapp/) and open console
4. Conduct a V4 typed data signature and verify recovery result
5. Return to add hardware flow and forget device
6. Go back to test dapp from full screen add hardware
7. Then open pop-up again by tapping on fox
8. Tap `Connect` to connect Account 1 or whatever account is now active after removing Ledger
9. Note error `Something went wrong, and we were unable to complete this action`
10. Tap `Dismiss`
11. Note console error in test dapp `MetaMask: 'eth_accounts' unexpectedly updated accounts. Please report this bug.`
12. `Connect` button in test dapp is no longer responsive, but `Use MetaMask` `Use Window.Ethereum` and `Wallet Connect` buttons do all remain interactive


<div style="display: inline-block"> </div>

Recording: https://recordit.co/bm4XY9l9QO

**Review by gantunesr — DISMISSED** [review]

**Review by k-g-j — DISMISSED** [review]

**Review by gantunesr — DISMISSED** [review]

**Review by gantunesr — DISMISSED** [review]

File: ui/selectors/selectors.js
Comment (danroc) [inline ui/selectors/selectors.js:367]: Can you add the documentation?

File: ui/selectors/selectors.js
Comment (danroc) [inline ui/selectors/selectors.js:373]: IMHO the `addressesOfKeyring` name is a bit confusing, since the scope of this variable is so small we could use just `addresses`.

**Review by k-g-j — APPROVED** [review]

**Review by gantunesr — APPROVED** [review]

File: ui/selectors/selectors.js
Comment (montelaidev) [reply ui/selectors/selectors.js:367]: Will address this in a small follow up pr

File: ui/selectors/selectors.js
Comment (montelaidev) [reply ui/selectors/selectors.js:373]: Will address this in a small follow up pr

**Comment by danroc** [comment]
My comments are to be addressed in a future PR.

File: ui/selectors/selectors.js
Comment (gantunesr) [reply ui/selectors/selectors.js:373]: Solved in https://github.com/MetaMask/metamask-extension/pull/23320

**Pull request merged**
